File: backend/src/main/java/de/knirpsenstadt/service/StatisticsService.java
package de.knirpsenstadt.service;

import de.knirpsenstadt.api.model.*;
import de.knirpsenstadt.model.Employee;
import de.knirpsenstadt.model.ScheduleEntry;
import de.knirpsenstadt.model.ScheduleEntryType;
import de.knirpsenstadt.model.TimeEntry;
import de.knirpsenstadt.repository.EmployeeRepository;
import de.knirpsenstadt.repository.GroupRepository;
import de.knirpsenstadt.repository.ScheduleEntryRepository;
import de.knirpsenstadt.repository.TimeEntryRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.time.DayOfWeek;
import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalTime;
import java.time.temporal.TemporalAdjusters;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class StatisticsService {

    private final EmployeeRepository employeeRepository;
    private final ScheduleEntryRepository scheduleEntryRepository;
    private final TimeEntryRepository timeEntryRepository;
    private final GroupRepository groupRepository;

    /**
     * Get overview statistics for a month
     */
    public OverviewStatistics getOverviewStatistics(LocalDate month) {
        LocalDate startOfMonth = month.withDayOfMonth(1);
        LocalDate endOfMonth = month.withDayOfMonth(month.lengthOfMonth());

        List<Employee> activeEmployees = employeeRepository.findAllActiveOrderByName();
        List<ScheduleEntry> scheduleEntries = scheduleEntryRepository.findByDateBetween(startOfMonth, endOfMonth);
        List<TimeEntry> timeEntries = timeEntryRepository.findByDateBetween(startOfMonth, endOfMonth);

        // Group by employee
        Map<Long, List<ScheduleEntry>> scheduleByEmployee = scheduleEntries.stream()
                .collect(Collectors.groupingBy(e -> e.getEmployee().getId()));
        Map<Long, List<TimeEntry>> timeByEmployee = timeEntries.stream()
                .collect(Collectors.groupingBy(e -> e.getEmployee().getId()));

        float totalScheduledHours = 0;
        float totalWorkedHours = 0;
        float totalOvertimeHours = 0;
        int sickDays = 0;
        int vacationDays = 0;

        List<EmployeeStatisticsSummary> employeeStats = new ArrayList<>();

        for (Employee employee : activeEmployees) {
            List<ScheduleEntry> empSchedule = scheduleByEmployee.getOrDefault(employee.getId(), List.of());
            List<TimeEntry> empTime = timeByEmployee.getOrDefault(employee.getId(), List.of());

            float scheduledHours = calculateScheduledHours(empSchedule);
            float workedHours = calculateWorkedHours(empTime);
            
            // Calculate expected hours for the month based on contract
            float contractedWeeklyHours = employee.getWeeklyHours() != null 
                    ? employee.getWeeklyHours().floatValue() : 0;
            int workingDaysInMonth = countWorkingDays(startOfMonth, endOfMonth);
            float expectedMonthlyHours = (contractedWeeklyHours / 5) * workingDaysInMonth;
            
            float overtimeHours = workedHours - expectedMonthlyHours;

            totalScheduledHours += scheduledHours;
            totalWorkedHours += workedHours;
            totalOvertimeHours += overtimeHours;

            // Count sick and vacation days
            for (ScheduleEntry entry : empSchedule) {
                if (entry.getEntryType() == ScheduleEntryType.SICK) {
                    sickDays++;
                } else if (entry.getEntryType() == ScheduleEntryType.VACATION) {
                    vacationDays++;
                }
            }

            EmployeeStatisticsSummary summary = new EmployeeStatisticsSummary();
            summary.setEmployee(AuthService.toApiEmployee(employee));
            summary.setScheduledHours(scheduledHours);
            summary.setWorkedHours(workedHours);
            summary.setOvertimeHours(overtimeHours);
            summary.setRemainingVacationDays(employee.getRemainingVacationDays() != null 
                    ? employee.getRemainingVacationDays().floatValue() : 0);
            employeeStats.add(summary);
        }

        OverviewStatistics stats = new OverviewStatistics();
        stats.setMonth(month);
        stats.setTotalEmployees(activeEmployees.size());
        stats.setEmployeeStats(employeeStats);
        stats.setTotalScheduledHours(totalScheduledHours);
        stats.setTotalWorkedHours(totalWorkedHours);
        stats.setTotalOvertimeHours(totalOvertimeHours);
        stats.setSickDays(sickDays);
        stats.setVacationDays(vacationDays);

        return stats;
    }

    /**
     * Get weekly statistics
     */
    public WeeklyStatistics getWeeklyStatistics(LocalDate weekStart) {
        LocalDate monday = weekStart.with(TemporalAdjusters.previousOrSame(DayOfWeek.MONDAY));
        LocalDate sunday = monday.plusDays(6);

        List<Employee> activeEmployees = employeeRepository.findAllActiveOrderByName();
        List<ScheduleEntry> scheduleEntries = scheduleEntryRepository.findByDateBetween(monday, sunday);
        List<TimeEntry> timeEntries = timeEntryRepository.findByDateBetween(monday, sunday);

        // Group by employee
        Map<Long, List<ScheduleEntry>> scheduleByEmployee = scheduleEntries.stream()
                .collect(Collectors.groupingBy(e -> e.getEmployee().getId()));
        Map<Long, List<TimeEntry>> timeByEmployee = timeEntries.stream()
                .collect(Collectors.groupingBy(e -> e.getEmployee().getId()));

        // Group by group
        Map<Long, List<ScheduleEntry>> scheduleByGroup = scheduleEntries.stream()
                .filter(e -> e.getGroup() != null)
                .collect(Collectors.groupingBy(e -> e.getGroup().getId()));

        float totalScheduledHours = 0;
        float totalWorkedHours = 0;

        List<EmployeeWeekSummary> byEmployee = new ArrayList<>();

        for (Employee employee : activeEmployees) {
            List<ScheduleEntry> empSchedule = scheduleByEmployee.getOrDefault(employee.getId(), List.of());
            List<TimeEntry> empTime = timeByEmployee.getOrDefault(employee.getId(), List.of());

            float scheduledHours = calculateScheduledHours(empSchedule);
            float workedHours = calculateWorkedHours(empTime);
            int daysWorked = (int) empTime.stream()
                    .filter(t -> t.getClockOut() != null)
                    .map(TimeEntry::getDate)
                    .distinct()
                    .count();

            totalScheduledHours += scheduledHours;
            totalWorkedHours += workedHours;

            EmployeeWeekSummary summary = new EmployeeWeekSummary();
            summary.setEmployee(toApiEmployeeWithWeeklyHours(employee));
            summary.setScheduledHours(scheduledHours);
            summary.setWorkedHours(workedHours);
            summary.setDaysWorked(daysWorked);
            byEmployee.add(summary);
        }

        // Calculate group summaries
        List<GroupWeekSummary> byGroup = new ArrayList<>();
        List<de.knirpsenstadt.model.Group> groups = groupRepository.findAllOrderByName();
        
        for (de.knirpsenstadt.model.Group group : groups) {
            List<ScheduleEntry> groupSchedule = scheduleByGroup.getOrDefault(group.getId(), List.of());
            float groupScheduledHours = calculateScheduledHours(groupSchedule);
            
            // Count staffed days (days with at least one schedule entry)
            int staffedDays = (int) groupSchedule.stream()
                    .map(ScheduleEntry::getDate)
                    .distinct()
                    .count();

            GroupWeekSummary groupSummary = new GroupWeekSummary();
            de.knirpsenstadt.api.model.Group groupDto = new de.knirpsenstadt.api.model.Group();
            groupDto.setId(group.getId());
            groupDto.setName(group.getName());
            groupDto.setColor(group.getColor());
            groupSummary.setGroup(groupDto);
            groupSummary.setTotalScheduledHours(groupScheduledHours);
            groupSummary.setStaffedDays(staffedDays);
            groupSummary.setUnderstaffedDays(5 - staffedDays); // Monday-Friday = 5 days
            byGroup.add(groupSummary);
        }

        WeeklyStatistics stats = new WeeklyStatistics();
        stats.setWeekStart(monday);
        stats.setWeekEnd(sunday);
        stats.setByEmployee(byEmployee);
        stats.setByGroup(byGroup);
        stats.setTotalScheduledHours(totalScheduledHours);
        stats.setTotalWorkedHours(totalWorkedHours);

        return stats;
    }

    /**
     * Get detailed statistics for a single employee
     */
    public EmployeeStatistics getEmployeeStatistics(Long employeeId, LocalDate month) {
        Employee employee = employeeRepository.findById(employeeId)
                .orElseThrow(() -> new RuntimeException("Employee not found"));

        LocalDate startOfMonth = month.withDayOfMonth(1);
        LocalDate endOfMonth = month.withDayOfMonth(month.lengthOfMonth());

        List<ScheduleEntry> scheduleEntries = scheduleEntryRepository
                .findByEmployeeIdAndDateBetween(employeeId, startOfMonth, endOfMonth);
        List<TimeEntry> timeEntries = timeEntryRepository
                .findByEmployeeIdAndDateBetween(employeeId, startOfMonth, endOfMonth);

        float contractedWeeklyHours = employee.getWeeklyHours() != null 
                ? employee.getWeeklyHours().floatValue() : 0;
        int workingDaysInMonth = countWorkingDays(startOfMonth, endOfMonth);
        float expectedMonthlyHours = (contractedWeeklyHours / 5) * workingDaysInMonth;

        float scheduledHours = calculateScheduledHours(scheduleEntries);
        float workedHours = calculateWorkedHours(timeEntries);
        float overtimeHours = workedHours - expectedMonthlyHours;

        int vacationDaysUsed = (int) scheduleEntries.stream()
                .filter(e -> e.getEntryType() == ScheduleEntryType.VACATION)
                .count();
        int sickDays = (int) scheduleEntries.stream()
                .filter(e -> e.getEntryType() == ScheduleEntryType.SICK)
                .count();

        // Build daily breakdown
        Map<LocalDate, List<ScheduleEntry>> scheduleByDate = scheduleEntries.stream()
                .collect(Collectors.groupingBy(ScheduleEntry::getDate));
        Map<LocalDate, List<TimeEntry>> timeByDate = timeEntries.stream()
                .collect(Collectors.groupingBy(TimeEntry::getDate));

        List<DayStatistics> dailyBreakdown = new ArrayList<>();
        LocalDate current = startOfMonth;
        while (!current.isAfter(endOfMonth)) {
            if (current.getDayOfWeek() != DayOfWeek.SATURDAY && current.getDayOfWeek() != DayOfWeek.SUNDAY) {
                List<ScheduleEntry> daySchedule = scheduleByDate.getOrDefault(current, List.of());
                List<TimeEntry> dayTime = timeByDate.getOrDefault(current, List.of());

                DayStatistics dayStats = new DayStatistics();
                dayStats.setDate(current);
                dayStats.setScheduledHours(calculateScheduledHours(daySchedule));
                dayStats.setWorkedHours(calculateWorkedHours(dayTime));
                
                if (!daySchedule.isEmpty()) {
                    dayStats.setEntryType(de.knirpsenstadt.api.model.ScheduleEntryType
                            .fromValue(daySchedule.get(0).getEntryType().name()));
                }
                
                dailyBreakdown.add(dayStats);
            }
            current = current.plusDays(1);
        }

        EmployeeStatistics stats = new EmployeeStatistics();
        stats.setEmployee(AuthService.toApiEmployee(employee));
        stats.setMonth(month);
        stats.setContractedHours(expectedMonthlyHours);
        stats.setScheduledHours(scheduledHours);
        stats.setWorkedHours(workedHours);
        stats.setOvertimeHours(overtimeHours);
        stats.setOvertimeBalance(employee.getOvertimeBalance() != null 
                ? employee.getOvertimeBalance().floatValue() : 0);
        stats.setVacationDaysUsed(vacationDaysUsed);
        stats.setVacationDaysRemaining(employee.getRemainingVacationDays() != null 
                ? employee.getRemainingVacationDays().floatValue() : 0);
        stats.setSickDays(sickDays);
        stats.setDailyBreakdown(dailyBreakdown);

        return stats;
    }

    /**
     * Calculate total scheduled hours from a list of schedule entries
     */
    private float calculateScheduledHours(List<ScheduleEntry> entries) {
        float totalMinutes = 0;
        for (ScheduleEntry entry : entries) {
            if (entry.getStartTime() != null && entry.getEndTime() != null 
                    && entry.getEntryType() == ScheduleEntryType.WORK) {
                long minutes = Duration.between(entry.getStartTime(), entry.getEndTime()).toMinutes();
                int breakMinutes = entry.getBreakMinutes() != null ? entry.getBreakMinutes() : 0;
                totalMinutes += (minutes - breakMinutes);
            }
        }
        return totalMinutes / 60f;
    }

    /**
     * Calculate total worked hours from a list of time entries
     */
    private float calculateWorkedHours(List<TimeEntry> entries) {
        float totalMinutes = 0;
        for (TimeEntry entry : entries) {
            if (entry.getClockIn() != null && entry.getClockOut() != null) {
                long minutes = Duration.between(entry.getClockIn(), entry.getClockOut()).toMinutes();
                int breakMinutes = entry.getBreakMinutes() != null ? entry.getBreakMinutes() : 0;
                totalMinutes += (minutes - breakMinutes);
            }
        }
        return totalMinutes / 60f;
    }

    /**
     * Count working days (Monday-Friday) between two dates
     */
    private int countWorkingDays(LocalDate start, LocalDate end) {
        int count = 0;
        LocalDate current = start;
        while (!current.isAfter(end)) {
            DayOfWeek day = current.getDayOfWeek();
            if (day != DayOfWeek.SATURDAY && day != DayOfWeek.SUNDAY) {
                count++;
            }
            current = current.plusDays(1);
        }
        return count;
    }

    /**
     * Convert employee to API model including weeklyHours
     */
    private de.knirpsenstadt.api.model.Employee toApiEmployeeWithWeeklyHours(Employee entity) {
        de.knirpsenstadt.api.model.Employee dto = AuthService.toApiEmployee(entity);
        // weeklyHours is already set by AuthService.toApiEmployee
        return dto;
    }
}

File: backend-management/internal/service/statistics_service.go
package service

import (
	"context"
	"fmt"
	"time"

	"github.com/knirpsenstadt/kita-apps/backend-management/internal/domain"
	"github.com/knirpsenstadt/kita-apps/backend-management/internal/repository"
)

// StatisticsService handles reporting.
type StatisticsService struct {
	employees   repository.EmployeeRepository
	schedules   repository.ScheduleRepository
	timeEntries repository.TimeEntryRepository
	groups      repository.GroupRepository
}

// NewStatisticsService creates a new StatisticsService.
func NewStatisticsService(
	employees repository.EmployeeRepository,
	schedules repository.ScheduleRepository,
	timeEntries repository.TimeEntryRepository,
	groups repository.GroupRepository,
) *StatisticsService {
	return &StatisticsService{
		employees:   employees,
		schedules:   schedules,
		timeEntries: timeEntries,
		groups:      groups,
	}
}

// OverviewStatistics represents the overview stats.
type OverviewStatistics struct {
	Month               time.Time
	TotalEmployees      int
	EmployeeStats       []EmployeeStatisticsSummary
	TotalScheduledHours float64
	TotalWorkedHours    float64
	TotalOvertimeHours  float64
	SickDays            int
	VacationDays        int
}

// EmployeeStatisticsSummary represents summary stats for an employee.
type EmployeeStatisticsSummary struct {
	Employee              domain.Employee
	ScheduledHours        float64
	WorkedHours           float64
	OvertimeHours         float64
	RemainingVacationDays float64
}

// EmployeeStatistics represents detailed employee stats.
type EmployeeStatistics struct {
	Employee              domain.Employee
	Month                 time.Time
	ContractedHours       float64
	ScheduledHours        float64
	WorkedHours           float64
	OvertimeHours         float64
	OvertimeBalance       float64
	VacationDaysUsed      int
	VacationDaysRemaining float64
	SickDays              int
	DailyBreakdown        []DayStatistics
}

// DayStatistics represents a daily breakdown.
type DayStatistics struct {
	Date           time.Time
	ScheduledHours float64
	WorkedHours    float64
	EntryType      *domain.ScheduleEntryType
}

// WeeklyStatistics represents weekly stats.
type WeeklyStatistics struct {
	WeekStart           time.Time
	WeekEnd             time.Time
	ByEmployee          []EmployeeWeekSummary
	ByGroup             []GroupWeekSummary
	TotalScheduledHours float64
	TotalWorkedHours    float64
}

// EmployeeWeekSummary represents weekly stats per employee.
type EmployeeWeekSummary struct {
	Employee       domain.Employee
	ScheduledHours float64
	WorkedHours    float64
	DaysWorked     int
}

// GroupWeekSummary represents weekly stats per group.
type GroupWeekSummary struct {
	Group               domain.Group
	TotalScheduledHours float64
	StaffedDays         int
	UnderstaffedDays    int
}

// Overview computes overview statistics for a month.
func (s *StatisticsService) Overview(ctx context.Context, month time.Time) (*OverviewStatistics, error) {
	startOfMonth := time.Date(month.Year(), month.Month(), 1, 0, 0, 0, 0, time.UTC)
	endOfMonth := startOfMonth.AddDate(0, 1, -1)

	employees, err := s.employees.List(ctx, true)
	if err != nil {
		return nil, err
	}

	scheduleEntries, err := s.schedules.List(ctx, startOfMonth, endOfMonth, nil, nil)
	if err != nil {
		return nil, err
	}

	timeEntries, err := s.timeEntries.List(ctx, startOfMonth, endOfMonth, nil)
	if err != nil {
		return nil, err
	}

	scheduleByEmployee := make(map[int64][]domain.ScheduleEntry)
	for _, entry := range scheduleEntries {
		scheduleByEmployee[entry.EmployeeID] = append(scheduleByEmployee[entry.EmployeeID], entry)
	}

	timeByEmployee := make(map[int64][]domain.TimeEntry)
	for _, entry := range timeEntries {
		timeByEmployee[entry.EmployeeID] = append(timeByEmployee[entry.EmployeeID], entry)
	}

	stats := &OverviewStatistics{
		Month:          startOfMonth,
		TotalEmployees: len(employees),
	}

	for _, employee := range employees {
		entries := scheduleByEmployee[employee.ID]
		times := timeByEmployee[employee.ID]

		scheduledHours := calculateScheduledHours(entries)
		workedHours := calculateWorkedHours(times)

		workingDays := countWorkingDays(startOfMonth, endOfMonth)
		expectedMonthlyHours := (employee.WeeklyHours / 5) * float64(workingDays)
		overtimeHours := workedHours - expectedMonthlyHours

		stats.TotalScheduledHours += scheduledHours
		stats.TotalWorkedHours += workedHours
		stats.TotalOvertimeHours += overtimeHours

		for _, entry := range entries {
			switch entry.EntryType {
			case domain.ScheduleEntryTypeSick:
				stats.SickDays++
			case domain.ScheduleEntryTypeVacation:
				stats.VacationDays++
			}
		}

		stats.EmployeeStats = append(stats.EmployeeStats, EmployeeStatisticsSummary{
			Employee:              employee,
			ScheduledHours:        scheduledHours,
			WorkedHours:           workedHours,
			OvertimeHours:         overtimeHours,
			RemainingVacationDays: employee.RemainingVacationDays,
		})
	}

	return stats, nil
}

// Weekly computes weekly statistics.
func (s *StatisticsService) Weekly(ctx context.Context, weekStart time.Time) (*WeeklyStatistics, error) {
	monday := weekStart
	for monday.Weekday() != time.Monday {
		monday = monday.AddDate(0, 0, -1)
	}
	sunday := monday.AddDate(0, 0, 6)

	employees, err := s.employees.List(ctx, true)
	if err != nil {
		return nil, err
	}

	scheduleEntries, err := s.schedules.List(ctx, monday, sunday, nil, nil)
	if err != nil {
		return nil, err
	}

	timeEntries, err := s.timeEntries.List(ctx, monday, sunday, nil)
	if err != nil {
		return nil, err
	}

	scheduleByEmployee := make(map[int64][]domain.ScheduleEntry)
	for _, entry := range scheduleEntries {
		scheduleByEmployee[entry.EmployeeID] = append(scheduleByEmployee[entry.EmployeeID], entry)
	}

	timeByEmployee := make(map[int64][]domain.TimeEntry)
	for _, entry := range timeEntries {
		timeByEmployee[entry.EmployeeID] = append(timeByEmployee[entry.EmployeeID], entry)
	}

	scheduleByGroup := make(map[int64][]domain.ScheduleEntry)
	for _, entry := range scheduleEntries {
		if entry.GroupID != nil {
			scheduleByGroup[*entry.GroupID] = append(scheduleByGroup[*entry.GroupID], entry)
		}
	}

	result := &WeeklyStatistics{
		WeekStart: monday,
		WeekEnd:   sunday,
	}

	for _, employee := range employees {
		entries := scheduleByEmployee[employee.ID]
		times := timeByEmployee[employee.ID]

		scheduledHours := calculateScheduledHours(entries)
		workedHours := calculateWorkedHours(times)

		daysWorked := countWorkedDays(times)

		result.TotalScheduledHours += scheduledHours
		result.TotalWorkedHours += workedHours

		result.ByEmployee = append(result.ByEmployee, EmployeeWeekSummary{
			Employee:       employee,
			ScheduledHours: scheduledHours,
			WorkedHours:    workedHours,
			DaysWorked:     daysWorked,
		})
	}

	groups, err := s.groups.List(ctx)
	if err != nil {
		return nil, err
	}

	for _, group := range groups {
		entries := scheduleByGroup[group.ID]
		groupHours := calculateScheduledHours(entries)
		staffedDays := countScheduledDays(entries)

		result.ByGroup = append(result.ByGroup, GroupWeekSummary{
			Group:               group,
			TotalScheduledHours: groupHours,
			StaffedDays:         staffedDays,
			UnderstaffedDays:    5 - staffedDays,
		})
	}

	return result, nil
}

// EmployeeStats computes statistics for a single employee.
func (s *StatisticsService) EmployeeStats(ctx context.Context, employeeID int64, month time.Time) (*EmployeeStatistics, error) {
	employee, err := s.employees.GetByID(ctx, employeeID)
	if err != nil {
		return nil, NewNotFound(fmt.Sprintf("Mitarbeiter mit ID %d nicht gefunden", employeeID))
	}

	startOfMonth := time.Date(month.Year(), month.Month(), 1, 0, 0, 0, 0, time.UTC)
	endOfMonth := startOfMonth.AddDate(0, 1, -1)

	scheduleEntries, err := s.schedules.List(ctx, startOfMonth, endOfMonth, &employeeID, nil)
	if err != nil {
		return nil, err
	}

	timeEntries, err := s.timeEntries.List(ctx, startOfMonth, endOfMonth, &employeeID)
	if err != nil {
		return nil, err
	}

	workingDays := countWorkingDays(startOfMonth, endOfMonth)
	expectedMonthlyHours := (employee.WeeklyHours / 5) * float64(workingDays)

	scheduledHours := calculateScheduledHours(scheduleEntries)
	workedHours := calculateWorkedHours(timeEntries)
	overtimeHours := workedHours - expectedMonthlyHours

	vacationDays := 0
	sickDays := 0
	for _, entry := range scheduleEntries {
		switch entry.EntryType {
		case domain.ScheduleEntryTypeVacation:
			vacationDays++
		case domain.ScheduleEntryTypeSick:
			sickDays++
		}
	}

	scheduleByDate := make(map[string][]domain.ScheduleEntry)
	for _, entry := range scheduleEntries {
		key := entry.Date.Format(dateLayout)
		scheduleByDate[key] = append(scheduleByDate[key], entry)
	}
	// time entries by date
	timeByDate := make(map[string][]domain.TimeEntry)
	for _, entry := range timeEntries {
		key := entry.Date.Format(dateLayout)
		timeByDate[key] = append(timeByDate[key], entry)
	}

	breakdown := make([]DayStatistics, 0)
	current := startOfMonth
	for !current.After(endOfMonth) {
		if current.Weekday() != time.Saturday && current.Weekday() != time.Sunday {
			key := current.Format(dateLayout)
			scheduleForDay := scheduleByDate[key]
			timeForDay := timeByDate[key]

			dayStats := DayStatistics{
				Date:           current,
				ScheduledHours: calculateScheduledHours(scheduleForDay),
				WorkedHours:    calculateWorkedHours(timeForDay),
			}
			if len(scheduleForDay) > 0 {
				entryType := scheduleForDay[0].EntryType
				dayStats.EntryType = &entryType
			}
			breakdown = append(breakdown, dayStats)
		}
		current = current.AddDate(0, 0, 1)
	}

	return &EmployeeStatistics{
		Employee:              *employee,
		Month:                 startOfMonth,
		ContractedHours:       expectedMonthlyHours,
		ScheduledHours:        scheduledHours,
		WorkedHours:           workedHours,
		OvertimeHours:         overtimeHours,
		OvertimeBalance:       employee.OvertimeBalance,
		VacationDaysUsed:      vacationDays,
		VacationDaysRemaining: employee.RemainingVacationDays,
		SickDays:              sickDays,
		DailyBreakdown:        breakdown,
	}, nil
}

func calculateScheduledHours(entries []domain.ScheduleEntry) float64 {
	minutes := 0.0
	for _, entry := range entries {
		if entry.EntryType != domain.ScheduleEntryTypeWork {
			continue
		}
		if entry.StartTime != nil && entry.EndTime != nil {
			diff := entry.EndTime.Sub(*entry.StartTime).Minutes()
			minutes += diff - float64(entry.BreakMinutes)
		}
	}
	return minutes / 60
}

func calculateWorkedHours(entries []domain.TimeEntry) float64 {
	minutes := 0.0
	for _, entry := range entries {
		if entry.ClockOut == nil {
			continue
		}
		diff := entry.ClockOut.Sub(entry.ClockIn).Minutes()
		minutes += diff - float64(entry.BreakMinutes)
	}
	return minutes / 60
}

func countWorkingDays(start, end time.Time) int {
	count := 0
	current := start
	for !current.After(end) {
		if current.Weekday() != time.Saturday && current.Weekday() != time.Sunday {
			count++
		}
		current = current.AddDate(0, 0, 1)
	}
	return count
}

func countWorkedDays(entries []domain.TimeEntry) int {
	unique := make(map[string]bool)
	for _, entry := range entries {
		if entry.ClockOut == nil {
			continue
		}
		unique[entry.Date.Format(dateLayout)] = true
	}
	return len(unique)
}

func countScheduledDays(entries []domain.ScheduleEntry) int {
	unique := make(map[string]bool)
	for _, entry := range entries {
		unique[entry.Date.Format(dateLayout)] = true
	}
	return len(unique)
}
